### DISCUSSION TITLE
Passing paramteres via the url?

### DISCUSSION THREAD

**Discussion by aashishpanta0** [jupyterlite/jupyterlite#1090] [Q&A] [answered]
I am trying to find a way to pass a parameter via jupyterlite URL like this: test.github.io/jplite-test/lab/index.html?name=dataset_id and be able to read it (dataset_id) from the notebooks inside the jupyterlite as a variable. Is there a way to do this?  Thanks!

**Comment by oscar6echo** [comment]
I am not aware that it is currently possible.

Maybe one way to do it is via a jupyterlab extension that would read the url upon change, extract query params (say starting with `kernel_` to filter them) and store it in the IndexedDB, accessible from Python kernels.

Would this avenue be feasible or is there a blocking obstable that I cannot see ?
Or a simpler way ?

**Comment by jtpio** [comment, marked as answer]
This is indeed not possible and will likely not be supported in JupyterLite by default.

However in JupyterLite it could make sense to work around some of this since everything runs in the browser. Maybe a good inspiration could be the `jupyterlab-open-url-parameter` extension as mentioned in https://jupyterlite.readthedocs.io/en/latest/howto/content/open-url-parameter.html. Then the params could indeed be stored in IndexedDB, or sent to the kernels on startup.

Another option (arguably simpler / cleaner) could be to create a [Jupyter Widget](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20Custom.html). The frontend part would be able to access the query parameters and sync them to the Python kernel.

**Comment by jsignell** [comment]
That suggestion was super helpful! 

I was able to implement the widget approach using [anywidget](https://github.com/manzt/anywidget). 

```python
import anywidget
import traitlets

from urllib import parse

class LocationWidget(anywidget.AnyWidget):
    _esm = """
    export function render({ model, el }) {
        model.set("location", String(window.location));
        model.save_changes();
    }
    """
    location = traitlets.Unicode().tag(sync=True)
    
    @property
    def query_params(self):
        return parse.parse_qs(parse.urlsplit(self.location).query)

location_widget = LocationWidget()
location_widget
```

This will render an empty output, but you do need to render it or the location value will not be set.

```python
location_widget.query_params
```

This seems to work fine in jupyterlite. It also works in voici with the caveat that it seems like I have to add a cell that gets `location_widget.location` _before_ getting `location_widget.query_params` I haven't quite been able to figure out why that might be.
